File: src/connectors/hunter.js
import { BaseConnector } from './base.js';
import https from 'https';

export class HunterConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'Hunter';
        this.apiKey = process.env.HUNTER_API_KEY;
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        if (!this.apiKey) return { status: 'skipped', error: 'HUNTER_API_KEY not set' };

        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];

            if (inputType === 'domain') {
                const data = await this._apiGet(`/v2/domain-search?domain=${encodeURIComponent(inputValue)}&api_key=${this.apiKey}&limit=50`);
                if (data?.data) {
                    const d = data.data;

                    if (d.organization) {
                        entities.push({
                            type: 'organization',
                            data: { name: d.organization, source: 'hunter' },
                            confidence: 0.8
                        });
                    }

                    for (const email of (d.emails || [])) {
                        entities.push({
                            type: 'email',
                            data: {
                                address: email.value,
                                type: email.type,
                                confidence_score: email.confidence,
                                first_name: email.first_name,
                                last_name: email.last_name,
                                position: email.position,
                                department: email.department,
                                source: 'hunter'
                            },
                            confidence: (email.confidence || 50) / 100,
                            asKnown: { type: 'email', value: email.value }
                        });

                        if (email.first_name && email.last_name) {
                            entities.push({
                                type: 'person',
                                data: {
                                    name: `${email.first_name} ${email.last_name}`,
                                    email: email.value,
                                    position: email.position,
                                    department: email.department,
                                    source: 'hunter'
                                },
                                confidence: 0.7
                            });
                        }
                    }
                }
            }

            if (inputType === 'email') {
                const data = await this._apiGet(`/v2/email-verifier?email=${encodeURIComponent(inputValue)}&api_key=${this.apiKey}`);
                if (data?.data) {
                    const d = data.data;
                    entities.push({
                        type: 'email',
                        data: {
                            address: inputValue,
                            status: d.status,
                            deliverable: d.result === 'deliverable',
                            disposable: d.disposable,
                            webmail: d.webmail,
                            mx_records: d.mx_records,
                            smtp_server: d.smtp_server,
                            source: 'hunter_verify'
                        },
                        confidence: d.score ? d.score / 100 : 0.7
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _apiGet(path) {
        return new Promise((resolve, reject) => {
            const req = https.get({
                hostname: 'api.hunter.io',
                path,
                headers: { 'Accept': 'application/json' },
                timeout: 15000
            }, res => {
                if (res.statusCode === 401) { res.resume(); return reject(new Error('Hunter API key invalid')); }
                if (res.statusCode === 429) { res.resume(); return reject(new Error('Hunter rate limited')); }
                if (res.statusCode >= 400) { res.resume(); return reject(new Error(`Hunter HTTP ${res.statusCode}`)); }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve(null); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('Hunter timeout')); });
        });
    }

    async healthCheck() {
        return { ok: !!this.apiKey, version: 'v2', hasApiKey: !!this.apiKey };
    }
}

File: cyclops/connectors/hunter.py
import os
import aiohttp
from cyclops.connectors.base import BaseConnector


class HunterConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "Hunter"
        self.api_key = os.environ.get("HUNTER_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        if not self.api_key:
            return {"status": "skipped", "error": "HUNTER_API_KEY not set"}
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            async with aiohttp.ClientSession() as session:
                if input_type == "domain":
                    async with session.get(f"https://api.hunter.io/v2/domain-search?domain={input_value}&api_key={self.api_key}", timeout=aiohttp.ClientTimeout(total=15)) as resp:
                        if resp.status >= 400:
                            raise Exception(f"Hunter HTTP {resp.status}")
                        data = await resp.json()
                    d = data.get("data", {})
                    if d.get("organization"):
                        entities.append({"type": "organization", "data": {"name": d["organization"], "source": "hunter"}, "confidence": 0.8})
                    for email_data in d.get("emails", []):
                        entities.append({"type": "email", "data": {
                            "address": email_data["value"], "type": email_data.get("type"),
                            "first_name": email_data.get("first_name"), "last_name": email_data.get("last_name"),
                            "position": email_data.get("position"), "department": email_data.get("department"),
                            "confidence": email_data.get("confidence"), "source": "hunter",
                        }, "confidence": (email_data.get("confidence", 50) / 100), "as_known": {"type": "email", "value": email_data["value"]}})
                        if email_data.get("first_name") and email_data.get("last_name"):
                            entities.append({"type": "person", "data": {"name": f"{email_data['first_name']} {email_data['last_name']}",
                                "position": email_data.get("position"), "source": "hunter"}, "confidence": 0.7})
                elif input_type == "email":
                    async with session.get(f"https://api.hunter.io/v2/email-verifier?email={input_value}&api_key={self.api_key}", timeout=aiohttp.ClientTimeout(total=15)) as resp:
                        if resp.status >= 400:
                            raise Exception(f"Hunter HTTP {resp.status}")
                        data = await resp.json()
                    d = data.get("data", {})
                    entities.append({"type": "email", "data": {"address": input_value, "result": d.get("result"), "score": d.get("score"),
                        "disposable": d.get("disposable"), "webmail": d.get("webmail"), "source": "hunter_verify"}, "confidence": 0.8})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": bool(self.api_key), "hasApiKey": bool(self.api_key)}
